data_model/src/block.rs

```diff
@@ -325,10 +313,88 @@ mod candidate {
         payload: BlockPayload,
     }
 
+    #[derive(Decode, Deserialize)]
+    struct BlockPayloadCandidate {
+        header: BlockHeader,
+        transactions: Vec<CommittedTransaction>,
+        event_recommendations: Vec<EventBox>,
+    }
+
+    #[derive(Decode, Deserialize)]
+    struct BlockHeaderCandidate {
+        height: NonZeroU64,
+        prev_block_hash: Option<HashOf<SignedBlock>>,
+        transactions_hash: HashOf<MerkleTree<SignedTransaction>>,
+        creation_time_ms: u64,
+        view_change_index: u32,
+        consensus_estimation_ms: u64,
+    }
+
+    impl BlockHeaderCandidate {
+        #[cfg(feature = "std")]
+        const fn creation_time(&self) -> Duration {
+            Duration::from_millis(self.creation_time_ms)
+        }
+
+        fn validate(self) -> Result<BlockHeader, &'static str> {
+            #[cfg(feature = "std")]
+            {
+                use std::time::SystemTime;
+
+                let now = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)
+                    .expect("INTERNAL BUG: Failed to get the current system time");
+
+                if now <= self.creation_time() {
+                    return Err("Block creation time is in the future");
+                }
```

Comment:
What is the reason for validating the upper and lower bounds in separate places?

Reply:
I prefer validating as soon as possible (i.e. on deserialization). But I can't validate lower bound on deserialization since I don't have access to the previous block

Reply:
However, validating on deserialization is redundant for clients. They should trust the integrity of the blockchain. In that case, I agree that all block check should be done only at the time of consensus. Never again later. I can make a separate ticket for that?

Reply:
#4950